manifests/01-validating-admission-policy.yaml

```
@@ -0,0 +1,36 @@
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: openshift-ingress-operator-gatewayapi-crd-admission
+  annotations:
+    capability.openshift.io/name: Ingress
+    include.release.openshift.io/ibm-cloud-managed: "true"
+    include.release.openshift.io/self-managed-high-availability: "true"
+    include.release.openshift.io/single-node-developer: "true"
+    release.openshift.io/feature-set: "TechPreviewNoUpgrade,CustomNoUpgrade"
+spec:
+  matchConstraints:
+    # Consider only requests to CRD resources.
+    resourceRules:
+    - apiGroups:   ["apiextensions.k8s.io"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE","UPDATE","DELETE"]
+      resources:   ["customresourcedefinitions"]
+  matchConditions:
+    # Consider only request to Gateway API CRDs.
+    - name: "check-only-gateway-api-crds"
+      # When the operation is DELETE, the "object" variable is null.
+      expression: "(request.operation == 'DELETE' ? oldObject : object).spec.group == 'gateway.networking.k8s.io'"
+  # Validations are evaluated in the the order of their declaration.
+  validations:
+    # Verify that the request was sent by the ingress operator's service account.
+    - expression: "has(request.userInfo.username) && (request.userInfo.username == 'system:serviceaccount:openshift-ingress-operator:ingress-operator')"
+      message: "modifications to Gateway API Custom Resource Definitions may only be made by the Ingress Operator"
+      reason: Forbidden
+    # Verify that the request was sent from a pod. The presence of both "node" and "pod" claims implies that the token is bound to a pod object.
+    # Ref: https://kubernetes.io/docs/reference/access-authn-authz/service-accounts-admin/#schema-for-service-account-private-claims.
+    - expression: "has(request.userInfo.extra) && ('authentication.kubernetes.io/node-name' in request.userInfo.extra) && ('authentication.kubernetes.io/pod-name' in request.userInfo.extra)"
+      message: "this user must have a \"authentication.kubernetes.io/node-name\" and \"authentication.kubernetes.io/pod-name\" claims"
```

nit, if we do need to check this:
```suggestion
      message: "unauthorized user"
```

---

I took this message pattern from other cluster oeprators: [ccm](https://github.com/openshift/cluster-cloud-controller-manager-operator/blob/e938398a53735d2f448fcf924b54f34b5b93af06/pkg/cloud/azure/assets/validating-admission-policy.yaml#L16) and [mco](https://github.com/openshift/machine-config-operator/blob/5dcc44abebbfb5dbeaaa8ef2569ba211032e1c00/manifests/machineconfigdaemon/mcn-guards-validatingadmissionpolicy.yaml#L20). I don't know to which degree we can diverge in this aspect across different cluster operators but I decided to stick to the same pattern for the sake of consistency.

---

I'm not sure if this is the error message that @danwinship said wasn't helpful, but to me it isn't helpful as is.  If you're reviewing the logs later, you might want to know the user name of the non-admin person trying to change CRDs.

---

> you might want to know the user name of the non-admin person trying to change CRDs.

Well, the user name cannot be anything but `system:serviceaccount:openshift-ingress-operator:ingress-operator` because the first validation ensures this. There can be [an impersonation involved](https://github.com/openshift/cluster-ingress-operator/pull/1192#discussion_r1987073723) but `request.userInfo.username` would not contain the name of the actual user anyway.

---

We'll go with the message as is.  We can follow up to find out whether it needs to follow a specific pattern or whether it can be made more user-friendly.

---

Okay.  I suggest changing the grammar a little too.
```suggestion
      message: "this user must have both \"authentication.kubernetes.io/node-name\" and \"authentication.kubernetes.io/pod-name\" claims"
```

---

Done.